NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,232 @@
+			sqvm, fmt::format("There was an error opening/creating file {} (Is the file name valid?)", fileName).c_str());
+		return SQRESULT_ERROR;
+	}
+	fileStr.write(content.c_str(), content.length());
+	fileStr.close();
+	return SQRESULT_NULL;
+}
+
+#include "scripts/scriptjson.h"
+
+ADD_SQFUNC("void", NSSaveJSONFile, "string file, table data", "", ScriptContext::CLIENT | ScriptContext::UI | ScriptContext::SERVER)
+{
+	Mod* mod = g_pSquirrel<context>->getcallingmod(sqvm);
+	if (mod == nullptr)
+	{
+		g_pSquirrel<context>->raiseerror(sqvm, "Has to be called from a mod function!");
+		return SQRESULT_ERROR;
+	}
+
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (CheckFileName(fileName))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format("File name invalid ({})! Make sure it has no '\\', '/' or non-ASCII charcters!", fileName, mod->Name).c_str());
+		return SQRESULT_ERROR;
+	}
+
+	std::string content = EncodeJSON<context>(sqvm);
+	if (ContainsNonASCIIChars(content))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm, fmt::format("File contents may not contain non-ASCII characters! Make sure your strings are valid!", mod->Name).c_str());
+		return SQRESULT_ERROR;
+	}
+
+	fs::create_directories(fs::path(GetNorthstarPrefix()) / "saveData" / fs::path(mod->m_ModDirectory).filename());
+	std::ofstream fileStr(fs::path(GetNorthstarPrefix()) / "saveData" / fs::path(mod->m_ModDirectory).filename() / (fileName));
+	if (fileStr.fail())
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm, fmt::format("There was an error opening/creating file {} (Is the file name valid?)", fileName).c_str());
+		return SQRESULT_ERROR;
+	}
+	fileStr.write(content.c_str(), content.length());
+	fileStr.close();
+
+	return SQRESULT_NULL;
+}
+
+ADD_SQFUNC("string", NSLoadFile, "string file", "", ScriptContext::CLIENT | ScriptContext::UI | ScriptContext::SERVER)
+{
+	Mod* mod = g_pSquirrel<context>->getcallingmod(sqvm);
+	if (mod == nullptr)
+	{
+		g_pSquirrel<context>->raiseerror(sqvm, "Has to be called from a mod function!");
+		return SQRESULT_ERROR;
+	}
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (CheckFileName(fileName))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format("File name invalid ({})! Make sure it has no '\\', '/' or non-ASCII charcters!", fileName, mod->Name).c_str());
+		return SQRESULT_ERROR;
+	}
+
+	std::ifstream fileStr(fs::path(GetNorthstarPrefix()) / "saveData" / fs::path(mod->m_ModDirectory).filename() / (fileName));
+	if (fileStr.fail())
+	{
+		g_pSquirrel<context>->pushstring(sqvm, "");
+		return SQRESULT_NOTNULL;
+	}
+
+	std::stringstream jsonStringStream;
+	while (fileStr.peek() != EOF)
+		jsonStringStream << (char)fileStr.get();
+	g_pSquirrel<context>->pushstring(sqvm, jsonStringStream.str().c_str());
+
+	return SQRESULT_NOTNULL;
+}
+
+ADD_SQFUNC("table", NSLoadJSONFile, "string file", "", ScriptContext::CLIENT | ScriptContext::UI | ScriptContext::SERVER)
```

Comment:
Ill mention this here just as a note.

Json files do not have to be an object (`table`).
They can be just an array, which we don't (and can't) handle atm because we return a `table` instead of a `var`

This usecase of JSON is perfectly valid, and Juicy (who I cant seem to @ ) has already ran into this problem grabbing data from speedrun.com

If anyone else is looking for the solution to this in the future, (in script) you can use string concatenation to encapsulate the JSON array in an object and then you can parse it into a table with the normal JSON functions
```
string jsonArray = "[\"foo\", \"bar\"]" // result from elsewhere, currently cannot be parsed into a table
jsonArray = "{ \"data\": " + jsonArray + "}" // encapsulate the array in an object
// the array data can now be parsed and used
```

This should be fixed, but should probably be done in a separate PR

Reply:
Might be able to return arrays, but encoding them will be harder to do.

For the sake of consistency, I'm keeping this table-only and resolving this, though if anyone has any idea on how to approach this. I'm down.

Reply:
Made into a new issue, can be resolved here just fine.